src/main/java/org/dependencytrack/persistence/MetricsQueryManager.java

```diff
@@ -62,33 +67,54 @@ public List<VulnerabilityMetrics> getVulnerabilityMetrics() {
         return execute(query).getList(VulnerabilityMetrics.class);
     }
 
+    /**
+     * Retrieve all the projects suitable for portfolio
+     */
+    private List<Project> getProjectsForPortfolio() {
```

Comment:
Did you do any testing of how this behaves when the principal has access to multiple thousands of projects? When I rewrote the metrics calculation in 4.6, I made sure to avoid `getAll` operations like the plague.

If you really *need* to fetch all projects at once, at the very least you should use a [fetch group](https://www.datanucleus.org/products/accessplatform_6_0/jdo/persistence.html#fetch_groups) to only retrieve the fields you need, which, looking at the code calling this method, would be only the ID or UUID.

There is a `METRICS_UPDATE` fetch group for `Project` already:

https://github.com/DependencyTrack/dependency-track/blob/7cfc466baf7548ddb758f9e7b008b2d4aaedd148/src/main/java/org/dependencytrack/model/Project.java#L92-L96

And here's how you would use it:

https://github.com/DependencyTrack/dependency-track/blob/7cfc466baf7548ddb758f9e7b008b2d4aaedd148/src/main/java/org/dependencytrack/tasks/metrics/PortfolioMetricsUpdateTask.java#L179

In any case, it would be helpful if you could do some profiling / sampling for this new functionality. I'm a fan of [VisualVM](https://visualvm.github.io/), but there are other alternatives that you could use.

Reply:
Thanks for suggestion @nscuro, will adapt to use a simplified fetch plan in this specific case.

Regarding the 1000+ projects using ACL, this seems super theoretical and highly unlikely. Still it could create somewhat useless pressure on the system, hence I suggest to simply return empty for anyone with above 100+ projects referenced through its ACL. This will solve the actual leak of information and it is still much better to have no data than data you should not see in the first place. Most users with ACL will not face this limitation. Will that work for you ?

Reply:
@mulder999 we have installations with 10K+ projects and one adopter that needs DT to be able to support 50K+ projects. If querying all projects is a requirement, then this feature needs to be disabled by default, have a configurable way to enable it in the UI, and provide a warning if they do.

Reply:
Hi @stevespringett, specific projects are fetched only for users that have ACL enable. Nothing has changed for those not using ACL.

Reply:
Right, but even for the installations that have ACLs enabled, this could be a big performance hit that they currently don't have today.

Reply:
@nscuro and @stevespringett I have implemented the changes requested in the backend. Please have a look. For sure teams with important/huge ACL would benefit of some additional work (namely some form of caching).

The web UI still requires a manual change to display the new config settings switch (would have dream of dynamic UI here but it did not happen) and the warning. Unsure when it will fit my bandwidth in the coming days.

Reply:
Web UI changes are implemented in frontend [PR 741](https://github.com/DependencyTrack/frontend/pull/741)